File: DSA#/Sorting-Algos/MergeSort.java
// Based on divide and conquer algorithm
// Divide the array into two halves and recursively sort them
// Then merge the arrays to get a sorted array

public class MergeSort{
    public static void mergeSort(int[] arr, int left, int right){
        if(left< right){
            int mid = left + (right-left) /2; //the middle element index
            mergeSort(arr, left, mid); //left subarray
            mergeSort(arr, mid+1, right); //right subarray
            merge(arr, left, mid, right);
        }
    }

    //to merge both sorted subarrays
    private static void merge(int[] arr, int left, int mid, int right){
        int n1= mid-left+1;
        int n2= right-mid;

        int[] leftArray = new int[n1];
        int[] rightArray = new int[n2];

        for(int i=0; i<n1; i++){
            leftArray[i] = arr[left+i];
        }

        for(int j=0; j <n2; j++){
            rightArray[j] = arr[mid+1+j];
        }

        int i=0;
        int j=0;
        int k=left;

        while(i<n1 && j<n2){
            if(leftArray[i]<=rightArray[j]){
                arr[k] = leftArray[i];
                i++;
            }
            else{
                arr[k] = rightArray[j];
                j++;
            }
            k++;
        }

        while(i<n1){
            arr[k] = leftArray[i];
            i++;
            k++;
        }

        while(j<n2){
            arr[k] = rightArray[j];
            j++;
            k++;
        }
    }

    public static void main(String[] args){
        int[] numbers = {5,2,3,12,5};
        mergeSort(numbers, 0, numbers.length-1);
        for(int num:numbers){
            System.out.print(num + " ");
        }
    }
}

File: typscript/merge_sort.js
/**
 * Merges two sorted subarrays into one sorted array.
 * @param left The left sorted array.
 * @param right The right sorted array.
 * @returns The merged and sorted array.
 */
function merge<T>(left: T[], right: T[]): T[] {
    const result: T[] = [];
    let i = 0; // Index for the left array
    let j = 0; // Index for the right array

    // Compare elements from both arrays and push the smaller one to the result
    while (i < left.length && j < right.length) {
        if (left[i] < right[j]) {
            result.push(left[i]);
            i++;
        } else {
            result.push(right[j]);
            j++;
        }
    }

    // Append any remaining elements (one of these loops will run)
    return result.concat(left.slice(i)).concat(right.slice(j));
}

/**
 * Sorts an array using the Merge Sort algorithm.
 * @param arr The array to sort.
 * @returns The fully sorted array.
 */
function mergeSort<T>(arr: T[]): T[] {
    // Base case: arrays with 0 or 1 element are already sorted
    if (arr.length <= 1) {
        return arr;
    }

    // 1. Divide the array into two halves (left and right)
    const mid = Math.floor(arr.length / 2);
    const left = arr.slice(0, mid);
    const right = arr.slice(mid);

    // 2. Recursively sort both halves
    const sortedLeft = mergeSort(left);
    const sortedRight = mergeSort(right);

    // 3. Merge the sorted halves
    return merge(sortedLeft, sortedRight);
}

// --- Usage Example ---
const unsortedArray = [38, 27, 43, 3, 9, 82, 10];
const sortedArray = mergeSort(unsortedArray);
console.log("\n--- Merge Sort ---");
console.log(`Unsorted: [${unsortedArray.join(', ')}]`);
console.log(`Sorted:   [${sortedArray.join(', ')}]`);
// Expected Output: [3, 9, 10, 27, 38, 43, 82]